### PR TITLE
Add DISC replay toolkit

### PR DESCRIPTION
**Pull request by Yancey1989**
related issue #76

### CODE REVIEW

File: tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      "p", llvm::cl::Required,
+      llvm::cl::desc("The tao_compiler_input protobuf message."),
+      llvm::cl::value_desc("program"), llvm::cl::cat(replay_cat)};
+  llvm::cl::opt<std::string> data_fname{
+      "d", llvm::cl::Required, llvm::cl::desc("The compressed input tensors."),
+      llvm::cl::value_desc("data"), llvm::cl::cat(replay_cat)};
+  llvm::cl::HideUnrelatedOptions(replay_cat);
+  llvm::cl::ParseCommandLineOptions(argc, argv, "Welcome BladeDISC!\n");
+
+  if (program_fname.empty() || data_fname.empty()) {
```
Comment (wyzero) [inline tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc:30]: maybe print a usage message here?

File: tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  }
+  auto record = replay::CreateReplayRecord(program_fname, data_fname);
+  if (record == nullptr) {
+    return tensorflow::errors::Internal("load replay record failed!");
+  }
+
+  replay::DiscInterpreter disc;
+  replay::CompiledResult result;
+  TF_RETURN_IF_ERROR(disc.Compile(record->Program(), result));
+  TF_RETURN_IF_ERROR(disc.Run(result, record->Tensors(), record->Placements()));
```
Comment (wyzero) [inline tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc:41]: nit: Add a flag to control warmup setting?

**Review by wyzero — COMMENTED** [review]
I think the overall structure is fine.  One thing maybe useful is to support fallback running as well. The fallback version can be used to do both performance and correctness check. We can leave it for the future.

File: tao_compiler/mlir/disc/tools/disc-replay/README.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+1. Find the input program and data file path from the training logs, the following is some log snippet
+
+    ``` text
+    ...
+    2022-02-09 09:22:54.244100: I /work/tao/tao_bridge/kernels/tao_compilation_cache.cc:792] tar the cluster input tensors with protobuf format to: /tmp/tempfile-4856de862901-217fa700-29301-5d79260e310f5.tar
+    ...
+    2022-02-09 09:22:55.011848: I /work/tao/tao_bridge/kernels/tao_compilation_cache.cc:817] tao_compiler_input: /tmp/tempfile-4856de862901-217fa700-29301-5d79260e40fdc.input
+   ```
+
+   The `tempfile-4856de862901-217fa700-29301-5d79260e310f5.tar` is a tarball which compressed input tensors, and `tempfile-4856de862901-217fa700-29301-5d79260e40fdc.input` is the compiler input program, the protobuf definition is
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/README.md:33]: "compressed" -> "compresses"

Comment (JamesTheZ) [reply tao_compiler/mlir/disc/tools/disc-replay/README.md:33]: NIT. Use separate sequences after `, and` and `, the`.

File: tao_compiler/mlir/disc/tools/disc-replay/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+   [tao_compiler_input.proto](/tao/tao_bridge/tao_compiler_input.proto).
+
+   These two files are all you need to replay a cluster.
+
+
+1. Replay the cluster using the `disc-replay-main` executable program, you can also use `nvprof` or other profiler toolkits
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/README.md:39]: 3.

Comment (JamesTheZ) [reply tao_compiler/mlir/disc/tools/disc-replay/README.md:39]: NIT. Separate the sequence into two.

File: tao_compiler/mlir/disc/tools/disc-replay/disc_interpreter.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+// function type for compilation executable file
+using func_t = void (*)(void**);
+
+struct CompiledResult {
+  // compiled result is a dynamic library file with suffix ".so"
+  std::string output_fname;
+  // meta file is a protobuf file with suffix ".pbtxt"
+  std::string meta_fname;
+};
+class DiscInterpreter {
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/disc_interpreter.h:44]: Better to add some comments to tell what this class does.

File: tao_compiler/mlir/disc/tools/disc-replay/record.cc [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    return "const";
+  } else if (arg_proto.kind_v2() ==
+             tensorflow::tao::ArgumentKind::kFixedShaped) {
+    return "cpu";
+  } else if (arg_proto.kind_v2() == tensorflow::tao::ArgumentKind::kHostArgs) {
+    return "cpu";
+  } else {
+    return default_device;
+  }
+  return "";
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/record.cc:31]: Unreachable code.

File: tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  tensorflow::TensorProto tensor_proto;
+  GetTestingTensor().AsProtoTensorContent(&tensor_proto);
+
+  // set argument kind
+  auto arg = input.add_args();
+  arg->set_kind_v2(tensorflow::tao::ArgumentKind::kConstant);
+
+  // set tensor value
+  std::string value_fn = "input_1.pb";
+  // std::string tao_input_pb_fn = "tao_compiler_input.pb";
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc:49]: NIT. Remove the unused comment.

File: tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  tensorflow::tao::TaoCompilerInput input;
+  tensorflow::TensorProto tensor_proto;
+  GetTestingTensor().AsProtoTensorContent(&tensor_proto);
+
+  // set argument kind
+  auto arg = input.add_args();
+  arg->set_kind_v2(tensorflow::tao::ArgumentKind::kConstant);
+
+  // set tensor value
+  std::string value_fn = "input_1.pb";
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc:48]: What if `input_1.pb` file already exists?

File: tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+TEST(BladeDISCReplayTest, TestRecordArgs) {
+  std::string tar_fname, program_fname;
+  EXPECT_TRUE(GetTestingRecord(tar_fname, program_fname).ok());
+
+  auto record = CreateReplayRecord(program_fname, tar_fname);
+  EXPECT_NE(nullptr, record);
+  auto tensors = record->Tensors();
+  // auto placements = record.Placements();
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc:72]: NIT. Unused comment.

File: tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+  // data value on the index 64(3 * 20 + 4) is equal to tensor[2][3]
+  float expected_value = 12;
+  EXPECT_EQ(tensors[0].flat<float>()(64), expected_value);
+}
+
+TEST(BladeDISCReplayTest, TestInterPreter_CUDA) {
+#ifndef GOOGLE_CUDA
+  GTEST_SKIP("skip for CPU build");
+#endif
+  // the test.tar is generate by quick start demo manually offline.
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/replay_test.cc:84]: NIT. "the" -> "The". "generate" -> "generated"

File: tao_compiler/mlir/disc/tools/disc-replay/tar_helper.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+tensorflow::Status DeCompressTar(const std::string& tar_fname,
+                                 const std::string& out_dir) {
+  // tar xf <tar_fname> -C <out_dir>
+  tensorflow::SubProcess proc;
+  std::string stdout;
+  std::string stderr;
+  std::vector<std::string> args = {"tar", "xf", tar_fname,
+                                   "-C " + out_dir + "/"};
+  proc.SetProgram("/bin/tar", {"tar", "xf", tar_fname, "-C", out_dir});
```
Comment (JamesTheZ) [inline tao_compiler/mlir/disc/tools/disc-replay/tar_helper.h:30]: NIT. Not sure whether `tar` is installed in popular OS by default.

File: tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc (hunk 1 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc]: I added the `Required` option on flag, that llvm Command Line would print some useful messages without required flags, and I also removed these two lines.

File: tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc (hunk 2 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/disc_replay_main.cc:62]: Good point, updated.

File: tao_compiler/mlir/disc/tools/disc-replay/README.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
+# BladeDISC Replay Toolkit
+
+In some scenarios, users want to profile or debug a single cluster using
+some exiting profiler toolkit such as `nvprof` or [Nvidia Nsight System](https://developer.nvidia.com/zh-cn/nsight-systems),
+this toolkit can help users to replay a single cluster on a host.
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/README.md:5]: done

File: tao_compiler/mlir/disc/tools/disc-replay/README.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+## How to Get the Toolkit
+
+An easy way to get the replay toolkit is by pulling the BladeDISC runtime Docker image, you can find the guide
+from [Install with Docker](/docs/install_with_docker.md).
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/README.md]: done

File: tao_compiler/mlir/disc/tools/disc-replay/README.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+## How to Use the Toolkit
+
+1. enable the debug mode on a TensorFlow script with the environment variable `DISC_DEBUG=true` as the following example
+
+    ``` bash
+    DISC_DEBUG=true python run.py 2>&1 > train.log
+    ```
+    
+1. Find the input program and data file path from the training logs, the following is some log snippet
```
[the remark this answers is not in the document]
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/README.md]: > "1. " -> "2. " ?

`1. .... 1.` is well, GitHub can show them as the consecutive serial numbers.

File: tao_compiler/mlir/disc/tools/disc-replay/README.md (hunk 3 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/README.md]: done

File: tao_compiler/mlir/disc/tools/disc-replay/disc_interpreter.h (hunk 5 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/disc_interpreter.h:54]: done

File: tao_compiler/mlir/disc/tools/disc-replay/record.cc (hunk 6 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/record.cc]: done

File: tao_compiler/mlir/disc/tools/disc-replay/tar_helper.h (hunk 11 above)
Comment (Yancey1989) [reply tao_compiler/mlir/disc/tools/disc-replay/tar_helper.h:30]: `tar` is installed in BladeDISC runtime Docker, and I also add a `CheckTarCommand` function to check the `tar` command.

**Review by wyzero — APPROVED** [review]
LGTM

**Comment by Yancey1989** [comment]
> I think the overall structure is fine. One thing maybe useful is to support fallback running as well. The fallback version can be used to do both performance and correctness check. We can leave it for the future.

Agree with that :)

**Pull request merged**